File: src/data/courses/dsa/merge-sort/merge-sort-applications/count-inversions/code/python/count_inversions.py
def merge_sort(arr, l, r):
    if l >= r:
        return 0
    m = l + (r - l) // 2
    return merge_sort(arr, l, m) + merge_sort(arr, m + 1, r) + merge(arr, l, m, r)

def merge(arr, l, m, r):
    L = arr[l:m+1]
    R = arr[m+1:r+1]
    i = j = 0
    k = l
    inv = 0
    while i < len(L) and j < len(R):
        if L[i] <= R[j]:
            arr[k] = L[i]
            i += 1
        else:
            arr[k] = R[j]
            j += 1
            inv += len(L) - i
        k += 1
    while i < len(L):
        arr[k] = L[i]
        i += 1
        k += 1
    while j < len(R):
        arr[k] = R[j]
        j += 1
        k += 1
    return inv

File: src/data/courses/dsa/merge-sort/merge-sort-applications/count-inversions/code/java/CountInversions.java
// Time: O(N log N), Space: O(N)

public class CountInversions {
    public static int countInversions(int[] arr) {
        int[] temp = new int[arr.length];
        return mergeSort(arr, temp, 0, arr.length - 1);
    }
    
    private static int mergeSort(int[] arr, int[] temp, int left, int right) {
        int count = 0;
        if (left < right) {
            int mid = left + (right - left) / 2;
            count += mergeSort(arr, temp, left, mid);
            count += mergeSort(arr, temp, mid + 1, right);
            count += merge(arr, temp, left, mid, right);
        }
        return count;
    }
    
    private static int merge(int[] arr, int[] temp, int left, int mid, int right) {
        int i = left, j = mid + 1, k = left, inversions = 0;
        
        while (i <= mid && j <= right) {
            if (arr[i] <= arr[j]) {
                temp[k++] = arr[i++];
            } else {
                temp[k++] = arr[j++];
                inversions += (mid - i + 1);
            }
        }
        
        while (i <= mid) temp[k++] = arr[i++];
        while (j <= right) temp[k++] = arr[j++];
        for (i = left; i <= right; i++) arr[i] = temp[i];
        
        return inversions;
    }
    
    public static void main(String[] args) {
        int[] arr = {8, 4, 2, 1};
        System.out.println("Inversions: " + countInversions(arr));
    }
}
